swarm/network/simulations/discovery/dynamic_test.go

```diff
@@ -0,0 +1,538 @@
+	self.readyNodes[nodeIdx] = self.readyNodes[len(self.readyNodes)-1]
+	self.readyNodes = self.readyNodes[:len(self.readyNodes)-1]
+
+	// find the selected node in the upNodes and upAddrs arrays (they have identical indices)
+	var found bool
+	for i, up := range self.upNodes {
+		if up == nodeId {
+			self.upNodes[i] = self.upNodes[len(self.upNodes)-1]
+			self.upAddrs[i] = self.upAddrs[len(self.upAddrs)-1]
+			self.upNodes = self.upNodes[:len(self.upNodes)-1]
+			self.upAddrs = self.upAddrs[:len(self.upAddrs)-1]
+			found = true
+			break
+		}
+	}
+
+	// this shouldn't happen
+	if !found {
+		self.mu.Unlock()
+		return fmt.Errorf("node %v listed as ready but not found in uplist", nodeId)
+	}
+
+	// stop the node
+	log.Info("Restart node: stop", "seq", seq, "id", nodeId, "addr", fmt.Sprintf("%x", self.upAddrs[nodeIdx]), "upnodes", len(self.upNodes), "readynodes", len(self.readyNodes))
+	self.pot = network.NewPeerPotMap(testMinProxBinSize, self.upAddrs)
+
+	if err := self.net.Stop(nodeId); err != nil {
+		self.mu.Unlock()
+		return err
+	}
+	self.mu.Unlock()
+
+	// wait a bit then bring back up
+	time.Sleep(self.randomDelay(0))
+
+	// start the node again
+	// wait for the up event and a conn event which involves the node
+	self.mu.Lock()
+	log.Info("Restart node: start", "seq", seq, "id", nodeId, "addr", fmt.Sprintf("%x", network.ToOverlayAddr(nodeId.Bytes())), "upnodes", len(self.upNodes), "readynodes", len(self.readyNodes))
+	self.mu.Unlock()
+
+	var err error
+	events := make(chan *simulations.Event)
+	sub := self.net.Events().Subscribe(events)
+	defer sub.Unsubscribe()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	errC := make(chan error)
+	quitC := make(chan struct{})
+	go func(errC chan error) {
+		for {
+			select {
+			case <-quitC:
+				return
+			case ev := <-events:
+				if ev == nil {
+					errC <- errors.New("got nil event")
+					return
+				} else if ev.Type == simulations.EventTypeNode {
+					if ev.Node.Up && ev.Node.Config.ID == nodeId {
+						log.Info("got node up event", "event", ev, "node", ev.Node.Config.ID)
+					}
+				} else if ev.Type == simulations.EventTypeConn {
+					if ev.Conn.Up && (ev.Conn.One == nodeId || ev.Conn.Other == nodeId) {
+						log.Info("got node conn event", "event", ev, "one", ev.Conn.One, "other", ev.Conn.Other)
+						errC <- nil
+					}
+				}
+
+			case <-ctx.Done():
+				errC <- ctx.Err()
+				return
+			}
+
+		}
+	}(errC)
+
+	if err := self.net.Start(nodeId); err != nil {
+		return err
+	}
+	if err = <-errC; err != nil {
+		return err
+	}
+	close(quitC)
+
+	sub.Unsubscribe()
+
+	// node is now up, we are connected to a peer
+	// we can add the node back to the uplist
+	self.mu.Lock()
+	self.upNodes = append(self.upNodes, nodeId)
+	self.upAddrs = append(self.upAddrs, network.ToOverlayAddr(nodeId.Bytes()))
+	self.mu.Unlock()
+
+	// then we poll for health
+	n := self.net.GetNode(nodeId)
+	err = self.checkHealth(n, seq)
```

Comment:
here - shouldn't all nodes be re-checked for health?

Reply:
see previous comment

Reply:
that is right, we need to check health of all nodes restored after a dropout

Reply:
That was not the question, I think. The question was if we need to check the health of all _remaining_ nodes when one node drops out.